Index: sys/arch/riscv/fdt/cpu_fdt.c
===================================================================
RCS file: /cvsroot/src/sys/arch/riscv/fdt/cpu_fdt.c,v
retrieving revision 1.4
diff -u -p -r1.4 cpu_fdt.c
--- sys/arch/riscv/fdt/cpu_fdt.c	1 Jan 2024 13:51:56 -0000	1.4
+++ sys/arch/riscv/fdt/cpu_fdt.c	16 Mar 2026 02:02:26 -0000
@@ -205,6 +205,7 @@ cpu_fdt_attach(device_t parent, device_t
 
 	/* Attach the CPU */
 	cpu_attach(self, hartid);
+	cpufunc_init(self, faa);
 
 	fdt_add_bus(self, phandle, faa);
 }
