Index: sys/arch/riscv/sunxi/files.sunxi
===================================================================
RCS file: /cvsroot/src/sys/arch/riscv/sunxi/files.sunxi,v
retrieving revision 1.5
diff -u -p -r1.5 files.sunxi
--- sys/arch/riscv/sunxi/files.sunxi	18 Mar 2026 06:42:35 -0000	1.5
+++ sys/arch/riscv/sunxi/files.sunxi	30 Mar 2026 08:02:43 -0000
@@ -45,6 +45,10 @@ device	sunxisid
 attach	sunxisid at fdt with sunxi_sid
 file	arch/arm/sunxi/sunxi_sid.c		sunxi_sid
 
+# USB OTG
+attach	motg at fdt with sunxi_musb: fdt_clock, fdt_phy, fdt_reset
+file	arch/arm/sunxi/sunxi_musb.c		sunxi_musb
+
 # USB PHY
 device	sunxiusbphy: fdt_phy, fdt_regulator, fdt_reset
 attach	sunxiusbphy at fdt with sunxi_usbphy
